flake.nix

```diff
@@ -1,66 +1,77 @@
+              nativeBuildInputs = old.nativeBuildInputs ++ [ pkgs.cmake ];
+              buildInputs = old.buildInputs ++ [ pkgs.scenefx ];
             });
       };
+
+      targetSystems = [
+        "aarch64-linux"
+        "x86_64-linux"
+      ];
+      pkgsFor =
+        system:
+        import nixpkgs {
+          inherit system;
+          overlays = [ scenefx.overlays.insert ];
+        };
+      forEachSystem = f: nixpkgs.lib.genAttrs targetSystems (system: f (pkgsFor system));
     in
     {
       overlays = rec {
         default = insert;
-        # Override onto the input nixpkgs
-        override = _: prev: mkPackage prev;
-        # Insert using the locked nixpkgs
+        # Insert using the locked nixpkgs. Can be used with any nixpkgs version.
         insert = _: prev: mkPackage (pkgsFor prev.system);
+        # Override onto the input nixpkgs. Users *MUST* have a scenefx overlay
+        # used before this overlay, otherwise pkgs.scenefx will be unavailable
+        override = _: prev: mkPackage prev;
       };
 
-      packages = nixpkgs.lib.genAttrs targetSystems (
-        system: (mkPackage (pkgsFor system) // { default = self.packages.${system}.swayfx-unwrapped; })
+      packages = forEachSystem (
+        pkgs: (mkPackage pkgs // { default = self.packages.${pkgs.system}.swayfx-unwrapped; })
       );
 
-      devShells = nixpkgs.lib.genAttrs targetSystems (
-        system:
-        let
-          pkgs = pkgsFor system;
-        in
-        {
-          default = pkgs.mkShell {
-            name = "swayfx-shell";
-            inputsFrom = [
-              self.packages.${system}.swayfx-unwrapped
-              pkgs.wlroots_0_17
-            ];
-            nativeBuildInputs = with pkgs; [
-              cmake
-              wayland-scanner
-              hwdata # for wlroots
-            ];
-            # Copy the nix version of wlroots into the project
-            shellHook = with pkgs; ''
-              (
-                mkdir -p "$PWD/subprojects" && cd "$PWD/subprojects"
-                cp -R --no-preserve=mode,ownership ${wlroots_0_17.src} wlroots
-              )'';
-          };
-        }
-      );
+      devShells = forEachSystem (pkgs: {
+        default = pkgs.mkShell {
+          name = "swayfx-shell";
+          inputsFrom = [
+            self.packages.${pkgs.system}.swayfx-unwrapped
+            pkgs.wlroots_0_17
+            pkgs.scenefx
+          ];
+          nativeBuildInputs = with pkgs; [
+            gdb # for debugging
+            wayland-scanner
+            hwdata # for wlroots
+          ];
+          # Copy the nix version of wlroots into the project
+          shellHook = ''
+            (
+              mkdir -p "$PWD/subprojects" && cd "$PWD/subprojects"
+              cp -R --no-preserve=mode,ownership ${pkgs.wlroots_0_17.src} wlroots
+              cp -R --no-preserve=mode,ownership ${pkgs.scenefx.src} scenefx
+            )'';
+        };
```

Comment:
Worth mentioning that if scenefx has wlroots as a subproject, swayfx will inherit that wlroots and won't need it to be added here. Not sure if that will happen when we switch to getting scenefx via the flake but if it does we could remove the wlroots line here

Reply:
By default the submodules aren't carried over i think, it's probably much easier to just keep things flat here

Reply:
Gotcha, wasn't sure if it's auto set up the scenefx dev env as well via the devshell if you use its flake, but it makes sense that it wouldn't